r/aws_apigatewayv2_stage: 'data_trace_enabled' and 'logging_level' are only valid for WebSocket APIs

ewbankkit opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes https://github.com/terraform-providers/terraform-provider-aws/issues/13801.
Relates https://github.com/terraform-providers/terraform-provider-aws/issues/13527.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
resource/aws_apigatewayv2_stage: Correctly handle `default_route_setting` and `route_setting` `data_trace_enabled` and `logging_level` for HTTP APIs. `logging_level` is now `Computed`, meaning Terraform will only perform drift detection of its value when present in a configuration.
```

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSAPIGatewayV2Stage_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSAPIGatewayV2Stage_ -timeout 120m
=== RUN   TestAccAWSAPIGatewayV2Stage_basicWebSocket
=== PAUSE TestAccAWSAPIGatewayV2Stage_basicWebSocket
=== RUN   TestAccAWSAPIGatewayV2Stage_basicHttp
=== PAUSE TestAccAWSAPIGatewayV2Stage_basicHttp
=== RUN   TestAccAWSAPIGatewayV2Stage_defaultHttpStage
=== PAUSE TestAccAWSAPIGatewayV2Stage_defaultHttpStage
=== RUN   TestAccAWSAPIGatewayV2Stage_disappears
=== PAUSE TestAccAWSAPIGatewayV2Stage_disappears
=== RUN   TestAccAWSAPIGatewayV2Stage_AccessLogSettings
=== PAUSE TestAccAWSAPIGatewayV2Stage_AccessLogSettings
=== RUN   TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription
=== PAUSE TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription
=== RUN   TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket
=== PAUSE TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket
=== RUN   TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp
=== PAUSE TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp
=== RUN   TestAccAWSAPIGatewayV2Stage_Deployment
=== PAUSE TestAccAWSAPIGatewayV2Stage_Deployment
=== RUN   TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket
=== PAUSE TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket
=== RUN   TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp
=== PAUSE TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp
=== RUN   TestAccAWSAPIGatewayV2Stage_StageVariables
=== PAUSE TestAccAWSAPIGatewayV2Stage_StageVariables
=== RUN   TestAccAWSAPIGatewayV2Stage_Tags
=== PAUSE TestAccAWSAPIGatewayV2Stage_Tags
=== CONT  TestAccAWSAPIGatewayV2Stage_basicWebSocket
=== CONT  TestAccAWSAPIGatewayV2Stage_Tags
=== CONT  TestAccAWSAPIGatewayV2Stage_StageVariables
=== CONT  TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp
=== CONT  TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket
=== CONT  TestAccAWSAPIGatewayV2Stage_Deployment
=== CONT  TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp
=== CONT  TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket
=== CONT  TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription
=== CONT  TestAccAWSAPIGatewayV2Stage_AccessLogSettings
=== CONT  TestAccAWSAPIGatewayV2Stage_disappears
=== CONT  TestAccAWSAPIGatewayV2Stage_defaultHttpStage
=== CONT  TestAccAWSAPIGatewayV2Stage_basicHttp
--- PASS: TestAccAWSAPIGatewayV2Stage_disappears (27.88s)
--- PASS: TestAccAWSAPIGatewayV2Stage_basicWebSocket (32.97s)
--- PASS: TestAccAWSAPIGatewayV2Stage_defaultHttpStage (33.62s)
--- PASS: TestAccAWSAPIGatewayV2Stage_basicHttp (33.94s)
--- PASS: TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket (51.24s)
--- PASS: TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp (51.90s)
--- PASS: TestAccAWSAPIGatewayV2Stage_StageVariables (52.18s)
--- PASS: TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket (52.24s)
--- PASS: TestAccAWSAPIGatewayV2Stage_Tags (52.35s)
--- PASS: TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp (52.43s)
--- PASS: TestAccAWSAPIGatewayV2Stage_Deployment (63.90s)
--- PASS: TestAccAWSAPIGatewayV2Stage_AccessLogSettings (73.62s)
--- PASS: TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription (108.73s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	108.789s
```

anGie44 commented on aws/resource_aws_apigatewayv2_stage.go:
```diff
@@ -99,7 +99,7 @@ func resourceAwsApiGatewayV2Stage() *schema.Resource {
 							DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
 								// Not set for HTTP APIs.
-								if old == "" && new == apigatewayv2.LoggingLevelOff {
+								if d.Id() != "" && old == "" && new == apigatewayv2.LoggingLevelOff {
```
could also use `d.IsNewResource()`

ewbankkit commented on aws/resource_aws_apigatewayv2_stage.go:
```diff
@@ -164,6 +164,13 @@ func resourceAwsApiGatewayV2Stage() *schema.Resource {
 							}, false),
+							DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+								// Not set for HTTP APIs.
+								if d.Id() != "" && old == "" && new == apigatewayv2.LoggingLevelOff {
```
Hmm. I added the `d.Id() != ""` so as to avoid changes from `"OFF"` to '""` in the WebSocket test 😄. I'll take another look at this.

anGie44 replied:
makes sense, thanks for taking a look! in general, it does seem tricky though since we're working around the `Default` in place.

ewbankkit replied:
Yes, if we could just change to no `Default` this wold be simpler.
The stumbling block now is the `route_settings` set hash function not understanding the suppressed fields...

reyescameron17 commented:
Any updates on when this might be merged? I'm assuming the PR has become stale due to a month passing since last commit.

ewbankkit commented:
@reyescameron17 Not gone stale, just swapped out for a bit 😄.
I am starting to think that changing these attributes to `Computed: true,` may be the way forward.

ewbankkit commented:
Rebased to fix merge conflicts and changed `logging_level` to `Compted: true`.
Acceptance test output:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSAPIGatewayV2Stage_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSAPIGatewayV2Stage_ -timeout 120m
=== RUN   TestAccAWSAPIGatewayV2Stage_basicWebSocket
=== PAUSE TestAccAWSAPIGatewayV2Stage_basicWebSocket
=== RUN   TestAccAWSAPIGatewayV2Stage_basicHttp
=== PAUSE TestAccAWSAPIGatewayV2Stage_basicHttp
=== RUN   TestAccAWSAPIGatewayV2Stage_defaultHttpStage
=== PAUSE TestAccAWSAPIGatewayV2Stage_defaultHttpStage
=== RUN   TestAccAWSAPIGatewayV2Stage_disappears
=== PAUSE TestAccAWSAPIGatewayV2Stage_disappears
=== RUN   TestAccAWSAPIGatewayV2Stage_AccessLogSettings
=== PAUSE TestAccAWSAPIGatewayV2Stage_AccessLogSettings
=== RUN   TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription
=== PAUSE TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription
=== RUN   TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket
=== PAUSE TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket
=== RUN   TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp
=== PAUSE TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp
=== RUN   TestAccAWSAPIGatewayV2Stage_Deployment
=== PAUSE TestAccAWSAPIGatewayV2Stage_Deployment
=== RUN   TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket
=== PAUSE TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket
=== RUN   TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp
=== PAUSE TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp
=== RUN   TestAccAWSAPIGatewayV2Stage_StageVariables
=== PAUSE TestAccAWSAPIGatewayV2Stage_StageVariables
=== RUN   TestAccAWSAPIGatewayV2Stage_Tags
=== PAUSE TestAccAWSAPIGatewayV2Stage_Tags
=== CONT  TestAccAWSAPIGatewayV2Stage_basicWebSocket
=== CONT  TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp
=== CONT  TestAccAWSAPIGatewayV2Stage_Tags
=== CONT  TestAccAWSAPIGatewayV2Stage_StageVariables
=== CONT  TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp
=== CONT  TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket
=== CONT  TestAccAWSAPIGatewayV2Stage_Deployment
=== CONT  TestAccAWSAPIGatewayV2Stage_AccessLogSettings
=== CONT  TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket
=== CONT  TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription
=== CONT  TestAccAWSAPIGatewayV2Stage_defaultHttpStage
=== CONT  TestAccAWSAPIGatewayV2Stage_disappears
=== CONT  TestAccAWSAPIGatewayV2Stage_basicHttp
--- PASS: TestAccAWSAPIGatewayV2Stage_disappears (16.92s)
--- PASS: TestAccAWSAPIGatewayV2Stage_basicHttp (20.92s)
--- PASS: TestAccAWSAPIGatewayV2Stage_defaultHttpStage (21.32s)
--- PASS: TestAccAWSAPIGatewayV2Stage_basicWebSocket (23.12s)
--- PASS: TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription (29.92s)
--- PASS: TestAccAWSAPIGatewayV2Stage_Tags (31.47s)
--- PASS: TestAccAWSAPIGatewayV2Stage_StageVariables (31.97s)
--- PASS: TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp (40.74s)
--- PASS: TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket (41.10s)
--- PASS: TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp (41.14s)
--- PASS: TestAccAWSAPIGatewayV2Stage_AccessLogSettings (52.82s)
--- PASS: TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket (67.46s)
--- PASS: TestAccAWSAPIGatewayV2Stage_Deployment (69.21s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	69.323s
```

ewbankkit commented:
@anGie44 The only solution here seems to be to change the `logging_level` attributes to `Computed` and note that in the documentation and release notes. I don't think this is a breaking change.

breathingdust commented:
LGTM 🚀 Thanks @ewbankkit 

Verified Acceptance Tests (Commercial)

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSAPIGatewayV2Stage_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSAPIGatewayV2Stage_ -timeout 120m
=== RUN   TestAccAWSAPIGatewayV2Stage_basicWebSocket
=== PAUSE TestAccAWSAPIGatewayV2Stage_basicWebSocket
=== RUN   TestAccAWSAPIGatewayV2Stage_basicHttp
=== PAUSE TestAccAWSAPIGatewayV2Stage_basicHttp
=== RUN   TestAccAWSAPIGatewayV2Stage_defaultHttpStage
=== PAUSE TestAccAWSAPIGatewayV2Stage_defaultHttpStage
=== RUN   TestAccAWSAPIGatewayV2Stage_disappears
=== PAUSE TestAccAWSAPIGatewayV2Stage_disappears
=== RUN   TestAccAWSAPIGatewayV2Stage_AccessLogSettings
=== PAUSE TestAccAWSAPIGatewayV2Stage_AccessLogSettings
=== RUN   TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription
=== PAUSE TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription
=== RUN   TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket
=== PAUSE TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket
=== RUN   TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp
=== PAUSE TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp
=== RUN   TestAccAWSAPIGatewayV2Stage_Deployment
=== PAUSE TestAccAWSAPIGatewayV2Stage_Deployment
=== RUN   TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket
=== PAUSE TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket
=== RUN   TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp
=== PAUSE TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp
=== RUN   TestAccAWSAPIGatewayV2Stage_StageVariables
=== PAUSE TestAccAWSAPIGatewayV2Stage_StageVariables
=== RUN   TestAccAWSAPIGatewayV2Stage_Tags
=== PAUSE TestAccAWSAPIGatewayV2Stage_Tags
=== CONT  TestAccAWSAPIGatewayV2Stage_basicWebSocket
=== CONT  TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp
=== CONT  TestAccAWSAPIGatewayV2Stage_Tags
=== CONT  TestAccAWSAPIGatewayV2Stage_StageVariables
=== CONT  TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp
=== CONT  TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket
=== CONT  TestAccAWSAPIGatewayV2Stage_Deployment
=== CONT  TestAccAWSAPIGatewayV2Stage_AccessLogSettings
=== CONT  TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket
=== CONT  TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription
=== CONT  TestAccAWSAPIGatewayV2Stage_defaultHttpStage
=== CONT  TestAccAWSAPIGatewayV2Stage_disappears
=== CONT  TestAccAWSAPIGatewayV2Stage_basicHttp
--- PASS: TestAccAWSAPIGatewayV2Stage_disappears (12.77s)
--- PASS: TestAccAWSAPIGatewayV2Stage_basicHttp (15.75s)
--- PASS: TestAccAWSAPIGatewayV2Stage_basicWebSocket (15.99s)
--- PASS: TestAccAWSAPIGatewayV2Stage_Deployment (19.15s)
--- PASS: TestAccAWSAPIGatewayV2Stage_defaultHttpStage (25.40s)
--- PASS: TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription (34.85s)
--- PASS: TestAccAWSAPIGatewayV2Stage_StageVariables (52.43s)
--- PASS: TestAccAWSAPIGatewayV2Stage_AccessLogSettings (54.68s)
--- PASS: TestAccAWSAPIGatewayV2Stage_Tags (55.09s)
--- PASS: TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket (61.99s)
--- PASS: TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp (64.04s)
--- PASS: TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket (71.80s)
--- PASS: TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp (87.28s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	90.736s
```

Verified Acceptance Tests (GovCloud)

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSAPIGatewayV2Stage_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSAPIGatewayV2Stage_ -timeout 120m
=== RUN   TestAccAWSAPIGatewayV2Stage_basicWebSocket
=== PAUSE TestAccAWSAPIGatewayV2Stage_basicWebSocket
=== RUN   TestAccAWSAPIGatewayV2Stage_basicHttp
=== PAUSE TestAccAWSAPIGatewayV2Stage_basicHttp
=== RUN   TestAccAWSAPIGatewayV2Stage_defaultHttpStage
=== PAUSE TestAccAWSAPIGatewayV2Stage_defaultHttpStage
=== RUN   TestAccAWSAPIGatewayV2Stage_disappears
=== PAUSE TestAccAWSAPIGatewayV2Stage_disappears
=== RUN   TestAccAWSAPIGatewayV2Stage_AccessLogSettings
=== PAUSE TestAccAWSAPIGatewayV2Stage_AccessLogSettings
=== RUN   TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription
=== PAUSE TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription
=== RUN   TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket
=== PAUSE TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket
=== RUN   TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp
=== PAUSE TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp
=== RUN   TestAccAWSAPIGatewayV2Stage_Deployment
=== PAUSE TestAccAWSAPIGatewayV2Stage_Deployment
=== RUN   TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket
=== PAUSE TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket
=== RUN   TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp
=== PAUSE TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp
=== RUN   TestAccAWSAPIGatewayV2Stage_StageVariables
=== PAUSE TestAccAWSAPIGatewayV2Stage_StageVariables
=== RUN   TestAccAWSAPIGatewayV2Stage_Tags
=== PAUSE TestAccAWSAPIGatewayV2Stage_Tags
=== CONT  TestAccAWSAPIGatewayV2Stage_basicWebSocket
=== CONT  TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp
=== CONT  TestAccAWSAPIGatewayV2Stage_AccessLogSettings
=== CONT  TestAccAWSAPIGatewayV2Stage_Deployment
=== CONT  TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription
=== CONT  TestAccAWSAPIGatewayV2Stage_disappears
=== CONT  TestAccAWSAPIGatewayV2Stage_basicHttp
=== CONT  TestAccAWSAPIGatewayV2Stage_StageVariables
=== CONT  TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp
=== CONT  TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket
=== CONT  TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket
=== CONT  TestAccAWSAPIGatewayV2Stage_defaultHttpStage
=== CONT  TestAccAWSAPIGatewayV2Stage_Tags
--- PASS: TestAccAWSAPIGatewayV2Stage_ClientCertificateIdAndDescription (19.78s)
--- PASS: TestAccAWSAPIGatewayV2Stage_StageVariables (22.20s)
--- PASS: TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsHttp (25.37s)
--- PASS: TestAccAWSAPIGatewayV2Stage_disappears (38.12s)
--- PASS: TestAccAWSAPIGatewayV2Stage_basicHttp (40.88s)
--- PASS: TestAccAWSAPIGatewayV2Stage_basicWebSocket (40.89s)
--- PASS: TestAccAWSAPIGatewayV2Stage_Deployment (44.25s)
--- PASS: TestAccAWSAPIGatewayV2Stage_Tags (44.86s)
--- PASS: TestAccAWSAPIGatewayV2Stage_AccessLogSettings (46.62s)
--- PASS: TestAccAWSAPIGatewayV2Stage_defaultHttpStage (47.80s)
--- PASS: TestAccAWSAPIGatewayV2Stage_RouteSettingsHttp (51.29s)
--- PASS: TestAccAWSAPIGatewayV2Stage_RouteSettingsWebSocket (59.57s)
--- PASS: TestAccAWSAPIGatewayV2Stage_DefaultRouteSettingsWebSocket (68.49s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	70.786s
```
